### PR TITLE
Allow CollectRnaSeqMetrics to take in a GTF file

### PR DESCRIPTION
**Pull request by sophiacrennan**
### Description

Most people use GTF files for RNA gene location information. CollectRNASeqMetrics tool accepts a refflat file which is not as common of a format anymore. It was requested to have the tool accept either a refflat or a GTF #805. 
CollectRNASeqMetrics now accepts either a refflat or a GTF file. If a GTF is accepted, then it is converted to a refflat using the class GtfToRefflat that I created. I have added unit tests with examples of GTF files and their respective refflat file.

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [ ] Added or modified tests to cover changes and any new functionality
- [ ] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

File: src/main/java/picard/analysis/GtfToRefflat.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+                    gene_id = feature.getID();
+                    chromosome = feature.getContig();
+                    strand = feature.getStrand();
+                    type = feature.getType().toLowerCase();
+
+                    if (type.equals("exon")) {
+                        exonStarts.add(feature.getStart() - 1);
+                        exonEnds.add(feature.getEnd());
+                        hasExon = true;
+                    } else if (hasExon == false) {
```
Comment (kishorikonwar) [inline src/main/java/picard/analysis/GtfToRefflat.java:95]: Can CDS be explicit?

Comment (gbggrant) [reply src/main/java/picard/analysis/GtfToRefflat.java]: @kishorikonwar I have inherited this PR from Sophie and don't know the domain that well.  Can you answer your own question here?  If you're not sure, I'm inclined to leave the code as is.  Thanks.

Comment (kishorikonwar) [reply src/main/java/picard/analysis/GtfToRefflat.java]: @gbggrant This is fine now and we have resolved it already, therefore, you can safely ignore it.

**Comment by gbggrant** [comment]
Assigning to @kachulis for questions about implementation detail

**Review by kachulis — CHANGES_REQUESTED** [review]
Haven't reviewed in detail yet, but after a quick initial look, it seems there is a lot of intermediate file creation being done to support reading in a new format.  I would propose instead a few different options:
1) Instead of making the tool work with GTF, just make it work with GFF3 instead.
2) If it is important to handle GTF, create a separate GTF codec, which can return either a new `GTFFeature` type or `Gff3Feature`. 
Whichever option is chosen, the main thing you need to extract from the input file is an `OverlapDetector<Gene>` object, which is then passed to RnaSeqMetricsCollector contstructor.  So all you really need is a `GtfReader` class to correspond to `RefFlatReader`.  That reader basically just needs to have a method to convert from a Gff3Feature to Gene, I think.

**Comment by gbggrant** [comment]
Putting this PR 'on hold'

Discussed with @kachulis  - he has concerns that the implementation itself may not be correct (so he is not okay with the PR being merged in its current state)

He would prefer that it be implemented using a specific codec for the GTC 'use a separate GTF codec, which can return either a new GTFFeature type or Gff3Feature'

He (Chris Kachulis) has a Picard branch in which he created a GFF3 codec, but it was not merged as he didn’t think it was feature complete: “the branch I had for this is ck_Gff3_RnaSeqMetrics .  It looks like I actually made an attempt at the ontology on that branch, which is probably a bit too far for now.  But you could rip that out and hard code in the feature types you care about.  And it looks like that branch still uses a local tag of htsjdk, but if you just rebase and bump the tag, it might just work now, or if not I don’t think it would take too much change to get it working.”

**Pull request closed without merging**